src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
Alright, in that case we can leave it. I generally prefer we opt for simplicity & correctness over tweakability. Grain Activations also live much longer than MVC Controllers tend to, so there's a difference there w.r.t cost of adding a scope (per-request vs (relatively) long-lived).